File: schema/type.cpp
#include <exception>
#include <capnp/schema.h>
#include "cgo.h"

pointer_result typeFromPrimitive(uint16_t which) {
	try {
		auto type = new capnp::Type;
		*type = capnp::Type(static_cast<capnp::schema::Type::Which>(which));
		return {static_cast<void*>(type), nullptr};
	} catch (const std::exception& e) {
		return {nullptr, strdup(e.what())};
	}
}

pointer_result typeFromStructSchema(void* structSchemaPtr) {
	try {
		auto schema = static_cast<capnp::StructSchema*>(structSchemaPtr);
		auto type = new capnp::Type;
		*type = capnp::Type(*schema);
		return {static_cast<void*>(type), nullptr};
	} catch (const std::exception& e) {
		return {nullptr, strdup(e.what())};
	}
}

pointer_result typeFromEnumSchema(void* enumSchemaPtr) {
	try {
		auto schema = static_cast<capnp::EnumSchema*>(enumSchemaPtr);
		auto type = new capnp::Type;
		*type = capnp::Type(*schema);
		return {static_cast<void*>(type), nullptr};
	} catch (const std::exception& e) {
		return {nullptr, strdup(e.what())};
	}
}

pointer_result typeFromInterfaceSchema(void* interfaceSchemaPtr) {
	try {
		auto schema = static_cast<capnp::InterfaceSchema*>(interfaceSchemaPtr);
		auto type = new capnp::Type;
		*type = capnp::Type(*schema);
		return {static_cast<void*>(type), nullptr};
	} catch (const std::exception& e) {
		return {nullptr, strdup(e.what())};
	}
}

pointer_result typeFromListSchema(void* listSchemaPtr) {
	try {
		auto schema = static_cast<capnp::ListSchema*>(listSchemaPtr);
		auto type = new capnp::Type;
		*type = capnp::Type(*schema);
		return {static_cast<void*>(type), nullptr};
	} catch (const std::exception& e) {
		return {nullptr, strdup(e.what())};
	}
}

uint16_t typeWhich(void* typePtr) {
	auto type = static_cast<capnp::Type*>(typePtr);
	return static_cast<uint16_t>(type->which());
}

pointer_result typeAsStruct(void* typePtr) {
	try {
		auto type = static_cast<capnp::Type*>(typePtr);
		auto newSchema = new capnp::StructSchema;
		*newSchema = type->asStruct();
		return {static_cast<void*>(newSchema), nullptr};
	} catch (const std::exception& e) {
		return {nullptr, strdup(e.what())};
	}
}

pointer_result typeAsEnum(void* typePtr) {
	try {
		auto type = static_cast<capnp::Type*>(typePtr);
		auto newSchema = new capnp::EnumSchema;
		*newSchema = type->asEnum();
		return {static_cast<void*>(newSchema), nullptr};
	} catch (const std::exception& e) {
		return {nullptr, strdup(e.what())};
	}
}

pointer_result typeAsInterface(void* typePtr) {
	try {
		auto type = static_cast<capnp::Type*>(typePtr);
		auto newSchema = new capnp::InterfaceSchema;
		*newSchema = type->asInterface();
		return {static_cast<void*>(newSchema), nullptr};
	} catch (const std::exception& e) {
		return {nullptr, strdup(e.what())};
	}
}

pointer_result typeAsList(void* typePtr) {
	try {
		auto type = static_cast<capnp::Type*>(typePtr);
		auto newSchema = new capnp::ListSchema;
		*newSchema = type->asList();
		return {static_cast<void*>(newSchema), nullptr};
	} catch (const std::exception& e) {
		return {nullptr, strdup(e.what())};
	}
}

void releaseType(void* typePtr) {
	auto type = static_cast<capnp::Type*>(typePtr);
	delete type;
}

File: schema/type.go
package schema

import (
	"runtime"
	"unsafe"
)

func newType(ptr unsafe.Pointer) *Type {
	t := &Type{
		ptr: ptr,
	}
	t.self = t
	runtime.SetFinalizer(t, (*Type).Release)
	return t
}

func TypeOf(primitive TypeWhich) *Type {
	return newType(mustPtr(typeFromPrimitive(uint16(primitive))))
}

type Type struct {
	self *Type
	ptr  unsafe.Pointer
}

func (t *Type) Which() TypeWhich {
	return TypeWhich(typeWhich(t.ptr))
}

func (t *Type) AsStruct() *Struct {
	return newStruct(mustPtr(typeAsStruct(t.ptr)))
}

func (t *Type) AsEnum() *Enum {
	return newEnum(mustPtr(typeAsEnum(t.ptr)))
}

func (t *Type) AsList() *List {
	return newList(mustPtr(typeAsList(t.ptr)))
}

func (t *Type) AsInterface() *Interface {
	return newInterface(mustPtr(typeAsInterface(t.ptr)))
}

func (t *Type) Release() {
	if t != t.self {
		panic("Schema should not be copied")
	}
	releaseType(t.ptr)
	t.ptr = nil
	runtime.SetFinalizer(t, nil)
}

type TypeWhich uint16

const (
	TypeVoid TypeWhich = iota
	TypeBool
	TypeInt8
	TypeInt16
	TypeInt32
	TypeInt64
	TypeUint8
	TypeUint16
	TypeUint32
	TypeUint64
	TypeFloat32
	TypeFloat64
	TypeText
	TypeData
	TypeList
	TypeEnum
	TypeStruct
	TypeInterface
	TypeAnyPointer
)

func (t TypeWhich) String() string {
	switch t {
	case TypeVoid:
		return "void"
	case TypeBool:
		return "bool"
	case TypeInt8:
		return "int8"
	case TypeInt16:
		return "int16"
	case TypeInt32:
		return "int32"
	case TypeInt64:
		return "int64"
	case TypeUint8:
		return "uint8"
	case TypeUint16:
		return "uint16"
	case TypeUint32:
		return "uint32"
	case TypeUint64:
		return "uint64"
	case TypeFloat32:
		return "float32"
	case TypeFloat64:
		return "float64"
	case TypeText:
		return "text"
	case TypeData:
		return "data"
	case TypeList:
		return "list"
	case TypeEnum:
		return "enum"
	case TypeStruct:
		return "struct"
	case TypeInterface:
		return "interface"
	case TypeAnyPointer:
		return "anyPointer"
	default:
		return ""
	}
}

func (t TypeWhich) IsPrimitive() bool {
	switch t {
	case TypeVoid,
		TypeBool,
		TypeInt8,
		TypeInt16,
		TypeInt32,
		TypeInt64,
		TypeUint8,
		TypeUint16,
		TypeUint32,
		TypeUint64,
		TypeFloat32,
		TypeFloat64,
		TypeText,
		TypeData:
		return true
	default:
		return false
	}
}
